### PR TITLE
Error reporting and apply() for File

### PR DESCRIPTION
**Pull request by Gondolav**
This PR adds a new `apply()` method for building templates out of `java.io.File` and error reporting for:

- YAML header syntax errors
- Non-existent top-level variable errors
- Non-existent non-top-level variable errors
- Non-existent function errors
- Wrong number of arguments for the function errors
- Invalid syntax errors
- Invalid function usage errors
- Invalid lambda usage errors

All tests have been update and enhanced to account for these new modifications.

### CODE REVIEW

File: thera/src/thera/Template.scala [hunk 1]
```diff
@@ -89,35 +97,75 @@ case class Template(argNames: List[String], context: ValueHierarchy, body: Body)
...
-        s"Variables outside function calls can only resolve to text. " +
-        s"Variable ${path.mkString(".")} was resolved to $x")
-      case x => x
+  private def evaluateNode(node: IndexedNode, inFunctionCall: Boolean)(
+      implicit ctx: ValueHierarchy): Value = {
+    val filename = fileInfo.file.value
+
+    def getLineColumnCode(name: String): (Int, Int, String) = {
+      if (fileInfo.isExternal) {
+        val lineCol = Future { indexToPosition(input, node.index) }
```
Comment (Gondolav) [inline thera/src/thera/Template.scala:106]: Remove Futures

File: thera/src/thera/Thera.scala [hunk 2]
```diff
@@ -1,12 +1,44 @@
...
+
+  private def buildTemplate(src: String, fileInfo: FileInfo): Template =
+    fastparse.parse(src, parser.module(src)(_, fileInfo)) match {
       case Success(result, _) => result
-      case f: Failure => throw new RuntimeException(f.toString)
+      case f: Failure =>
+        val code = getCodeSnippetFromParsingFailure(f)
+
+        val ((line, lineNb), column) = {
+          val ln = Future {
```
Comment (Gondolav) [inline thera/src/thera/Thera.scala:31]: Remove Futures

File: thera/src/thera/Thera.scala [hunk 3]
```diff
@@ -1,12 +1,44 @@
...
+            getLine(code, fileInfo.file.value)
+          }
+          val col = Future {
+            getColumnFromParsingFailure(f)
+          }
+          (Await.result(ln, Duration.Inf), Await.result(col, Duration.Inf))
+        }
+
+        throw if (isLambda(code)) EvaluationError(fileInfo.file.value, lineNb, column, line, InvalidLambdaUsageError)
```
Comment (Gondolav) [inline thera/src/thera/Thera.scala:40]: Extract filename

File: thera/src/thera/Value.scala [hunk 4]
```diff
@@ -88,7 +116,10 @@ trait ValueHierarchy extends Value {
     }
 
   final def apply(path: List[String]): Value = get(path) match {
-    case null => throw new RuntimeException(s"Value not found: ${path.mkString(".")}")
+    case null =>
+      val name = path.mkString(".")
+      if (path.length == 1) throw InternalParserError(NonExistentFunctionError(name))
```
Comment (Gondolav) [inline thera/src/thera/Value.scala:121]: Remove it

File: thera/src/thera/Value.scala [hunk 5]
```diff
@@ -128,16 +159,18 @@ object ValueHierarchy {
...
       def searchInMapping(path: List[String], m: JsonObject): Value =
         path match {
           case Nil => null
           case name :: Nil => valueFromNode(m(name).orNull)
-          case name :: rest => m(name).orNull match {
-            case x if x.isObject => searchInMapping(rest, x.asObject.orNull)
-            case _ => null
+          case name :: rest =>
+            m(name).orNull match {
+              case null => throw InternalParserError(NonExistentTopLevelVariableError(path.mkString(".")))
```
Comment (Gondolav) [inline thera/src/thera/Value.scala:171]: InternalEvaluationError instead

File: thera/src/thera/reporting/ErrorType.scala [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+// Parser errors
+
+sealed trait ParserErrorType extends ErrorType
+
+case object YamlError extends ParserErrorType {
+  override def toErrorMessage: String = "Syntax error in YAML header"
+}
+
+sealed abstract class NonExistentVariableError(val variable: String) extends ParserErrorType
```
Comment (Gondolav) [inline thera/src/thera/reporting/ErrorType.scala:15]: Change them to EvaluationErrorType

File: thera/src/thera/reporting/Utils.scala [hunk 7]
```diff
@@ -0,0 +1,47 @@
+package thera.reporting
+
+import fastparse.Parsed.Failure
+
+import scala.io.Source
+import scala.util.Using
+
+object Utils {
+
+  private[thera] def getLineNb(s: String, filename: String): Int = getLine(s, filename)._2
```
Comment (Gondolav) [inline thera/src/thera/reporting/Utils.scala:10]: Use public instead

Comment (Gondolav) [reply thera/src/thera/reporting/Utils.scala:10]: Dotty dropped private[this], not what we use

Comment (anatoliykmetyuk) [reply thera/src/thera/reporting/Utils.scala:10]: Oh, ok, so we can leave this here.

**Review by anatoliykmetyuk — APPROVED** [review]

**Comment by anatoliykmetyuk** [comment]
Nice work, thank you @Gondolav 🎉

**Pull request merged**